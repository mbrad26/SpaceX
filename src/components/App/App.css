body {
  font-family: 'Space Mono', monospace;
  /* background-color: black; */
  background-color: #e5e5e5;
  /* background: url(./falcon-9-rocket-4928x3264-spacex-cape-canaveral-4k-6186.jpg) no-repeat center center fixed;
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover; */
}

#title {
  padding: 20px;
  width: 1000px;
}

.container {
  padding: 20px;
  width: 850px;
}

/* Navbar */

#navbar {
  z-index: 1;
}

nav {
  margin: 20px 75px;
}

.links-container {
  padding: 0px;
  margin: 0px;
}

/*
img {
  display: block;
  width: 100%;
  height: auto;
} */

btn {
  border-style: hidden;
  border: none;
}

/* List and Item */

.row {
  padding: 20px;
}

.card-title {
  font-size: 1.55rem;
  /* font-weight: normal; */
}

.item-description {
  font-size: 0.9rem;
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 5;
  -webkit-box-orient: vertical;
}

.card-img-top {
  width: 100%;
  height: 35vh;
}

/* Modal */

Modal .container-fluid {
  padding: 16px;
}

#wiki {
  margin-bottom: 0px;
}

.row {
  padding: 0px 0px;

}

.col {
  padding:  0px;
  font-size: 0.9rem;
  justify-content: center;
}

.carousel-item {
  width: 100%;
  height: 375px;
  overflow: hidden;
}
.carousel-item img {
  width: 100%;
  height: 375px;
  object-fit: cover;
}
