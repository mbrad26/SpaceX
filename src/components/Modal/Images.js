import React from 'react';

import { Image, Carousel } from 'react-bootstrap';

const Images = ({ activeItem }) => (
  <Carousel>
    {activeItem.flickr_images.map(url =>
      <Carousel.Item key={url}>
        <Image
          src={url}
          alt='First slide'
        />
      </Carousel.Item>
    )}
  </Carousel>
)

export default Images;
