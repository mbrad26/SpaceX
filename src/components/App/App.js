import React, {
  useEffect,
  useReducer,
  useCallback,
  useRef,
} from 'react';
import List from '../List/List';
import NavBar from '../NavBar/NavBar';
import ModalComponent from '../Modal/Modal';
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import axios from 'axios';
import './App.css';

const Context = React.createContext(null);

const API_ENDPOINT = 'https://api.spacexdata.com/v3/';

const dataReducer = (state, action) => {
  switch (action.type) {
    case 'DATA_LOADING':
      return {
        ...state,
        loading: true,
        isError: false,
      };
    case 'DATA_SUCCESS':
      return {
        ...state,
        loading: false,
        data: action.payload,
      };
    case 'DATA_ERROR':
      return {
        ...state,
        loading: false,
        isError: true,
      }
    case 'SET_URL':
      return {
        ...state,
        url: `${API_ENDPOINT}${action.payload}`,
      }
    case 'OPEN_MODAL':
      return {
        ...state,
        isOpen: true,
        activeItem: action.payload,
      }
    case 'CLOSE_MODAL':
      return {
        ...state,
        isOpen: false,
      }
    default:
      throw new Error();
  }
};

const App = () => {
  const initialState = {
    data: [],
    loading: false,
    isError: false,
    url: '',
    isOpen: false,
    activeItem: '',
  };
  const [state, dispatch] = useReducer(dataReducer, initialState);
  const isMounted = useRef(false);

  const fetchData = useCallback(async () => {
    dispatch({ type: 'DATA_LOADING' });

    try {
      const result = await axios.get(state.url);
      dispatch({
        type: 'DATA_SUCCESS',
        payload: result.data
      });
    } catch {
        dispatch({ type: 'DATA_ERROR' });
    }
  }, [state.url]);

  const handleClick = event => {
    dispatch({ type: 'SET_URL', payload: event.target.innerHTML });
  }

  const handleOpenModal = item => {
    dispatch({ type: 'OPEN_MODAL', payload: item});
  }

  const handleCloseModal = () => {
    dispatch({ type: 'CLOSE_MODAL' });
  }

  useEffect(() => {
    if(!isMounted.current) {
      isMounted.current = true;
    } else {
      fetchData();
    }
  }, [fetchData]);

  return (
    <Context.Provider value={{
        ...state,
        handleClick: handleClick,
        handleOpenModal: handleOpenModal,
        handleCloseModal: handleCloseModal,
        dispatch: dispatch,
      }}
      >
      <NavBar />
      <Container id="title">
        <Row className='justify-content-center'>
          <Col lg={12}>
            <h2>
              “You want to wake up in the morning and think the future is going
              to be great - and that’s what being a spacefaring civilization is all
              about. It’s about believing in the future and thinking that the future
              will be better than the past. And I can’t think of anything more
              exciting than going out there and being among the stars.”
            </h2>
            <h2 className="text-right">
              -Elon Musk
            </h2>
          </Col>
        </Row>
      </Container>
      <Container>
        <Row>
          {state.isError && <h3>Something is wrong ...</h3>}

          {state.loading ? <p>Loading ...</p> : <List />}

          {state.isOpen && <ModalComponent />}
        </Row>
      </Container>
    </Context.Provider>
  )
};

export default App;
export { Context, dataReducer };
