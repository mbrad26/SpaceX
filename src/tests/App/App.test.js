import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import { itemOne, itemTwo, dragon, dataRockets } from '../fixtures.js';
import App, { Context, dataReducer } from '../../components/App/App.js';
import ModalComponent from '../../components/Modal/Modal';
import Images from '../../components/Modal/Images.js';
import axios from 'axios';

jest.mock('axios');

const path =  "https://api.spacexdata.com/v3/"
const endPoint = 'rockets';

describe('App', () => {
  it('renders snapshot', () => {
    const { container } = render(<App />);

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('#handleCloseModal', () => {
    it('closes a modal', async () => {
      const dispatch = jest.fn();
      const context = {
        handleCloseModal: jest.fn(),
        activeItem: itemOne,
        isOpen: true,
      }
      render(
          <Context.Provider value={context}>
            {context.isOpen && <ModalComponent />}
          </Context.Provider>
      )

      fireEvent.click(screen.getAllByRole('button', { name: 'Close' })[1]);
    });
  });

  describe('#fetchData', () => {
    it('fetches data', async () => {
      const result = Promise.resolve({
          data: dataRockets,
      });
      axios.get.mockImplementationOnce(() => result);

      render(<App />)

      expect(screen.queryByText(/Loading/)).not.toBeInTheDocument();

      const rockets = screen.getByText('ROCKETS');
      fireEvent.click(rockets);

      expect(screen.queryByText(/Loading/)).toBeInTheDocument();

      await act(() => result);

      expect(screen.queryByText(/Loading/)).toBeNull();
      expect(screen.getByText('Rocket One')).toBeInTheDocument();
      expect(screen.getByText('Rocket Two')).toBeInTheDocument();
      expect(screen.getByText(itemOne.description)).toBeInTheDocument();
      expect(screen.getByText(itemTwo.description)).toBeInTheDocument();
    });

    it('fails to fetch data', async () => {
      const reject = Promise.reject();
      axios.get.mockImplementationOnce(() => reject);
      render(<App />);

      fireEvent.click(screen.getByText('ROCKETS'));

      expect(screen.queryByText('Loading ...')).toBeInTheDocument();

      try {
        await act(() => reject);
      } catch {
        expect(screen.queryByText('Something is wrong ...')).toBeInTheDocument();
        expect(screen.queryByText('Loading ...')).not.toBeInTheDocument();
      }
    });
  });

  describe('#dataReducer', () => {
    it('returns loading == true', () => {
      const action = { type: 'DATA_LOADING' };
      const state = {
        data: [],
        loading: false,
        isError: false,
      }

      const newState = dataReducer(state, action);
      const expectedState = { data: [], loading: true, isError: false}

      expect(newState).toStrictEqual(expectedState);
    });

    it('returns data on success', () => {
      const action = { type: 'DATA_SUCCESS', payload: dataRockets }
      const state = {
        data: [],
        loading: true,
        isError: false
      }

      const newState = dataReducer(state, action);
      const expectedState = { data: dataRockets, loading: false, isError: false };

      expect(newState).toStrictEqual(expectedState);
    });

    it('returns isError on failure', () => {
      const action = { type: 'DATA_ERROR' }
      const state = {
        data: [],
        loading: true,
        isError: false
      }

      const newState = dataReducer(state, action);
      const expectedState = { data: [], loading: false, isError: true };

      expect(newState).toStrictEqual(expectedState);
    });

    it('sets URL', () => {
      const action = { type: 'SET_URL', payload:  endPoint }
      const state = {
        url: ''
      }

      const newState = dataReducer(state, action);
      const expectedState = { url: `${path}${endPoint}` };

      expect(newState).toStrictEqual(expectedState);
    });

    it('opens the modal', () => {
      const action = { type: 'OPEN_MODAL', payload:  itemOne }
      const state = {
        isOpen: false,
        activeItem: ''
      }

      const newState = dataReducer(state, action);
      const expectedState = { isOpen: true, activeItem: itemOne };

      expect(newState).toStrictEqual(expectedState);
    });

    it('closes the modal', () => {
      const action = { type: 'CLOSE_MODAL' }
      const state = {
        isOpen: true,
      }

      const newState = dataReducer(state, action);
      const expectedState = { isOpen: false };

      expect(newState).toStrictEqual(expectedState);
    });

    it('throws a new error', () => {
      const action = { type: '' };
      const state = {};

      expect(() => dataReducer(state, action)).toThrow(new Error());
    });
  });
});
