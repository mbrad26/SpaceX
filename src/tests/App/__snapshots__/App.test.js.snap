// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders snapshot 1`] = `
<nav
  class="navbar navbar-expand-lg navbar-light"
>
  <div
    class="links-container container"
  >
    <a
      href="#"
    >
      <img
        alt=""
        class="d-inline-block align-top"
        loading="lazy"
        src="SpaceX-Logo.wine.svg"
      />
    </a>
    <ul
      class="navbar-nav mr-auto"
    >
      <li
        class="nav-item active"
      >
        <a
          class="nav-link"
          href="#"
        >
          ROCKETS
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          class="nav-link"
          href="#"
        >
          DRAGONS
        </a>
      </li>
    </ul>
  </div>
</nav>
`;
